### PR TITLE
Replace `x509-certificate` with `x509-cert`

### LINKED ISSUE #20: Dependency `x509-certificate` has copy-left MPL-2.0 license
This just came up in my automatic license check using `cargo-deny` and I wanted to raise awareness, because in my project and company this is a problem.

#14 brought in the crate `x509-certificate`, which has the copy-left MPL-2.0 license. `tokio-postgres-rustls` is pretty fundamental infrastructure: when a project uses Postgres and Rust, the chances are good, that this crate is used. Keeping the licenses permissive would allow it to be used in environments that ban the use of copy-left licenses.

Would this crate be open to exchanging that dependency?

In my naive view, `rustls` should have all the x509 infrastructure included, because `rustls` also needs certificate parsing. It would be great to work with the already existing dependencies. But I also read the docs and have not yet found how to get the required information.

### PR DESCRIPTION
**Pull request by aumetra**
Closes #20

### CODE REVIEW

**Review by serprex — APPROVED** [review]

File: src/lib.rs
Comment (serprex) [inline src/lib.rs:109]:
```suggestion
                    match cert.signature.oid {
                        // Note: SHA1 is upgraded to SHA256 as per https://datatracker.ietf.org/doc/html/rfc5929#section-4.1
                        ID_SHA_1
                        | ID_SHA_256
                        | SHA_1_WITH_RSA_ENCRYPTION
                        | SHA_256_WITH_RSA_ENCRYPTION
                        | ECDSA_WITH_SHA_256 => &digest::SHA256,
                        ID_SHA_384 | SHA_384_WITH_RSA_ENCRYPTION | ECDSA_WITH_SHA_384 => {
                            Some(&digest::SHA384)
                        }
                        ID_SHA_512 | SHA_512_WITH_RSA_ENCRYPTION | ID_ED_25519 => Some(&digest::SHA512),
                        _ => None,
                    }
```

**Review by brooksmtownsend — APPROVED** [review]
Hey! Of course I'm not a maintainer on this repo/crate, but I am a maintainer of the CNCF project wasmCloud. We use this crate as a part of our PostgreSQL provider, and the license of the x509-certificate is preventing us from turning TLS on 😢 

Just wanted to hopefully nudge this along!

**Comment by brooksmtownsend** [comment]
@aumetra / @serprex just bumping for awareness

**Comment by serprex** [comment]
Neither of us have rights on this repo, it's up to @jbg to merge this PR

File: src/lib.rs
Comment (aumetra) [reply src/lib.rs:109]: We actually have two cases with `Some`, that's why I chose the structure to return `None` and put the others into a variable.

**Comment by aumetra** [comment]
So, at this point I decided to fork the crate and continue its development under `tokio-postgres-rustls2`.
Hopefully I can sync it back to upstream at some point.

Gonna continue the versioning where this crate took off. Kinda. First version will be `0.13.0`, currently got a release candidate, gonna try to get the second open PR integrated in some form.

<https://crates.io/crates/tokio-postgres-rustls2>

**Comment by aumetra** [comment]
Whoops. Accidentally closed because I changed stuff with the repo. Will reopen with a full merge of all upstream changes if desired

**Comment by serprex** [comment]
> Whoops. Accidentally closed because I changed stuff with the repo. Will reopen with a full merge of all upstream changes if desired

Keeping this open would be good. Will your fork include #25?

**Comment by aumetra** [comment]
I don't like the solution proposed by the PR, but I want to include something like that in my fork, yes

**Comment by jbg** [comment]
Apologies for not getting to this earlier, the recent comments bumped it back up my inbox though. I'd like to understand the reason for replacing the dependency. The MPL 2.0 does not prevent usage in commercial software, and does not require publishing source. You just have to publish source _if you change x509-certificate_. I feel like this change is being made because of misunderstanding of the license which doesn't seem like a good justification to make a change.

**Comment by aumetra** [comment]
For me it's less about the license and more because I prefer RustCrypto crates.

I might be biased due to some involvement with the project, but I personally feel like their code is more consistently high quality.

Plus for my dependency tree at least, it had some reduction in dependency count. But again, that's all very subjective from my perspective.

**Comment by jbg** [comment]
OK, if `x509-cert` is higher quality / has a smaller footprint then that seems a perfectly decent reason for the change. I see @brooksmtownsend mentioned that `x509-certificate`'s MPL 2.0 license prevents their organisation from using it due to their policy, which I guess is a problem even if I personally think that policy is overcautious or based on a misunderstanding of the license. So, I will have a quick look over `x509-cert` and then I think there's no problem with merging this.

**Comment by aumetra** [comment]
By the way, I'd be happy to support you in maintaining this crate since I depend on it for a project of mine.

When we merge this, I would also put in a few other PRs that get rid of a few allocations, and make most of the API private.

Also a slightly better solution to supporting multiple crypto backends than proposed by #25

**Comment by jbg** [comment]
Just for clarity, we already support multiple `rustls` crypto backends, and don't force any particular one to be chosen (except in tests where we have to pick one to run the tests with). #25 is just about the crate that is used for digests for doing the pg channel binding. TBH, I'd be tempted to just use some pure-Rust digest crate from RustCrypto rather than make it switchable between two crates that do a _lot_ more than digests.

**Comment by joonas** [comment]
I'll echo @aumetra's sentiment about preference for the Rust Crypto ecosystem, but also note that from a CNCF project perspective, MPL 2.0 isn't on the [list of allowed licenses](https://github.com/cncf/foundation/blob/main/allowed-third-party-license-policy.md#approved-licenses-for-allowlist), and while there's a process for requesting exceptions, it's quite toilsome and drawn out, which definitely incentives using licenses that align with CNCF's guidance.

**Comment by jbg** [comment]
@aumetra I'm unable to re-open this as the source repository has been deleted — are you able to reinstate the source repository (or file a new PR if that's not possible)?

**Comment by aumetra** [comment]
Will do as soon as I'm at a computer

File: Cargo.toml
Comment (jbg) [inline Cargo.toml:12]: Can you add `default-features = false` here as well? This protects against `const-oid` adding default features in a minor (or even patch) release, which isn't considered semver-breaking if nothing was removed or changed in the API, but would pull in additional dependencies that we might not need.

Comment (aumetra) [reply Cargo.toml:12]: Done in the PR I opened.

**Comment by aumetra** [comment]
Donezo. Opened a PR that addresses your review

**Pull request closed without merging**
